# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [2]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [3]:
KSUExpend = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [4]:
KSURev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [5]:
KUExpend = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [6]:
KURev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [7]:
KSURev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

In [8]:
KURev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1077 entries, 0 to 1076
Data columns (total 22 columns):
Fiscal_Year                        1077 non-null int64
Fiscal_Quarter                     1077 non-null int64
Business_Unit                      1077 non-null int64
Business_Unit_Descr                1077 non-null object
GAAP_Fund_Type                     1077 non-null int64
GAAP_Fund_Type_Descr               1077 non-null object
Fund_Code                          1077 non-null int64
Fund_Code_Descr                    1077 non-null object
Budget_Ref                         1077 non-null int64
Budget_Ref_Descr                   1077 non-null object
Program_Code                       1077 non-null int64
Program_Code_Descr                 1077 non-null object
State_Function                     1077 non-null int64
State_Function_Descr               1077 non-null object
Primary_Account_Code               1077 non-null int64
Primary_Account_Code_Descr         1077 non-null object
Interm

In [21]:
KSURev.Total.sum()

854324643.84

In [22]:
KURev.Total.sum()

808767902.55

In [23]:
854324643.84-808767902.55

45556741.29000008

Answer: At K-State there was 854,324,643.84 and KU 808,767,902.55. KSU had 45,556,741.29 more.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [24]:
KSUExpend.Total.sum()

1018584373.02

In [25]:
KUExpend.Total.sum()

990238875.6099999

In [26]:
1018584373.02 - 990238875.6099999

28345497.410000086

<b>Answer:</b> KSU 1,018,584,373.02 KU 990,238,875.6099999. 28,345,497.410000086 more at KSU

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [33]:
KSURev.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [34]:
KURev.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

<b>Answer: Agenecy Earnings was the primary source with 637 at K-State and 376 at KU

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [35]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [39]:
KURev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

In [41]:
KSURev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [42]:
412565112.820-408409337.320

4155775.5

Answer: Agency earnings was the biggest source at $412,565,112.820 at KSU, $408,409,337.320 at KU

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [2]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

In [59]:
KSURev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                412565112.820
GIFT DONATION FEDERAL GRANTS   278778848.030
NON REVENUE RECEIPTS            71947564.050
OTHER REVENUE                    8432030.160
REV FROM USE OF MONEY & PROP    34139318.040
REVENUE TRANSFERS               48458616.490
TAXES                               3154.250
Name: Total, dtype: float64

In [60]:
KURev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

<b>Answer:</b> Agency earnings was the biggest source of revenue for the universities 

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

<b>Answer:</b> 

### Q7: What costs two universities the most? 

In [50]:
KUExpend.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    70676545.250
COMMODITIES                       24225857.120
CONTRACTUAL SERVICES             139630507.780
DEBT SERVICE                      25309742.030
EXPENSE TRANSFERS                 25815417.520
GRANTS CLAIMS SHARED REVENUE      71439705.360
NONEXPENSE ITEMS                 161145084.400
SALARIES AND WAGES               471996016.150
Name: Total, dtype: float64

In [51]:
KSUExpend.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    64350550.130
COMMODITIES                       42670908.840
CONTRACTUAL SERVICES             128541598.280
DEBT SERVICE                      32442445.100
EXPENSE TRANSFERS                 37130576.230
GRANTS CLAIMS SHARED REVENUE     106551248.640
NONEXPENSE ITEMS                 130529992.790
SALARIES AND WAGES               476367053.010
Name: Total, dtype: float64

<b>Answer:</b> Salaries and wages cost the universities the most

### Q8: Calculate the percentage of this cost over each university's total expense.

In [52]:
KSUExpend.Total.sum()

1018584373.02

In [54]:
476367053.010 / 1018584373.02 * 100

46.76755952946929

In [55]:
KUExpend.Total.sum()

990238875.6099999

In [56]:
471996016.150 / 990238875.6099999 * 100

47.66486428431163

<b>Answer: At K-State it was 46.76755952946929% perecent of the total expense, KU it was 47.66486428431163% of the total expenses

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [46]:
854324643.84-1018584373.02

-164259729.17999995

In [48]:
808767902.55-990238875.6099999

-181470973.05999994

Answer: Neither univeristy made a profit in 2018

### B2: Which software program do you like the most in this course? Why?

Answer: I enjoyed Tableau becuase I felt like it was the easiest to understand. I also made some pretty cool graphics on it. The maps we made about the crime were really cool. 

### END OF THE TEST